File: dji_link_beta/composite.py
"""
DUSS composite-mux demux for the raw AOA stream (confirmed by reversing:
duss_parse_composite_data @0x491a070, thread mb_route_usb_data_recv_task).

Raw bulk from AOA = a stream of units:
  [0]=0x55 [1]=0xCC | type u16 LE | length u32 LE | payload[length]
  full unit size = 8 + length

Routing by type (index = type - 0x5749):
  0x5749 -> DUML channel  (payload = standard DUML frame, starts with 0x55)
  0x574A / 0x574D -> video (payload starts with 0x6d, 16-byte liveview header)
  0x574B / 0x574C / 0x7530 -> other channels (ignored for now)

A unit can be split across USB reads -> we buffer; resync on 0x55 0xCC.
"""

from __future__ import annotations
import struct

SOF0, SOF1 = 0x55, 0xCC
TYPE_DUML = 0x5749
TYPE_VIDEO = (0x574A, 0x574D)


def wrap(payload: bytes, typ: int = TYPE_DUML) -> bytes:
    """Wrap a payload (e.g. a DUML frame) into a composite unit for sending over AOA.
    The Pi simply forwards this to the RC without parsing anything (jump-host)."""
    return bytes([SOF0, SOF1]) + struct.pack("<H", typ) + struct.pack("<I", len(payload)) + payload


class CompositeDemux:
    def __init__(self, on_duml=None, on_video=None, max_unit=0x200000):
        self.buf = bytearray()
        self.on_duml = on_duml       # callback(payload: bytes) — DUML frame(s)
        self.on_video = on_video     # callback(payload: bytes) — starts with 0x6d
        self.max_unit = max_unit
        self.units = 0

    def feed(self, data: bytes) -> None:
        self.buf += data
        while True:
            if len(self.buf) < 8:
                break
            if not (self.buf[0] == SOF0 and self.buf[1] == SOF1):
                # resync: look for the next 55 CC
                idx = self.buf.find(bytes([SOF0, SOF1]), 1)
                if idx < 0:
                    keep = 1 if self.buf[-1] == SOF0 else 0   # partial SOF
                    del self.buf[:len(self.buf) - keep]
                    break
                del self.buf[:idx]
                continue
            typ = struct.unpack_from("<H", self.buf, 2)[0]
            length = struct.unpack_from("<I", self.buf, 4)[0]
            if length > self.max_unit:
                del self.buf[0]                 # garbage — shift
                continue
            total = 8 + length
            if len(self.buf) < total:
                break                            # wait for the rest (split between reads)
            payload = bytes(self.buf[8:total])
            del self.buf[:total]
            self.units += 1
            if typ == TYPE_DUML and self.on_duml:
                self.on_duml(payload)
            elif typ in TYPE_VIDEO and self.on_video:
                self.on_video(payload)
            # other types — skip


def feed_video_payload(reassembler, payload: bytes) -> None:
    """video unit payload = [16-byte LvHeader][H.264 slice] -> into the reassembler."""
    from liveview import LvHeader, HDR_LEN
    if len(payload) < HDR_LEN or payload[0] != 0x6D:
        return
    hdr = LvHeader.parse(payload[:HDR_LEN])
    reassembler.feed_packet(hdr, payload[HDR_LEN:HDR_LEN + hdr.pkt_len])

File: src/core/composite.cpp
#include "core/composite.hpp"

namespace djilink {
namespace {
constexpr std::uint16_t kVideoA = 0x574A;
constexpr std::uint16_t kVideoD = 0x574D;
} // namespace

Bytes composite_wrap(const Bytes& payload, std::uint16_t typ) {
    Bytes out;
    out.reserve(8 + payload.size());
    out.push_back(COMPOSITE_SOF0);
    out.push_back(COMPOSITE_SOF1);
    put_u16(out, typ);
    put_u32(out, static_cast<std::uint32_t>(payload.size()));
    out.insert(out.end(), payload.begin(), payload.end());
    return out;
}

CompositeDemux::CompositeDemux(DumlCb on_duml, VideoCb on_video, UnitCb on_unit,
                               std::size_t max_unit)
    : on_duml_(std::move(on_duml)), on_video_(std::move(on_video)), on_unit_(std::move(on_unit)),
      max_unit_(max_unit) {}

void CompositeDemux::feed(const std::uint8_t* data, std::size_t len) {
    buf_.insert(buf_.end(), data, data + len);
    while (true) {
        if (buf_.size() < 8)
            break;
        if (!(buf_[0] == COMPOSITE_SOF0 && buf_[1] == COMPOSITE_SOF1)) {
            // resync: look for the next 55 CC starting at index 1
            std::size_t idx = std::string::npos;
            for (std::size_t i = 1; i + 1 < buf_.size(); ++i) {
                if (buf_[i] == COMPOSITE_SOF0 && buf_[i + 1] == COMPOSITE_SOF1) {
                    idx = i;
                    break;
                }
            }
            if (idx == std::string::npos) {
                const std::size_t keep = (buf_.back() == COMPOSITE_SOF0) ? 1 : 0; // partial SOF
                buf_.erase(buf_.begin(), buf_.end() - static_cast<std::ptrdiff_t>(keep));
                break;
            }
            buf_.erase(buf_.begin(), buf_.begin() + static_cast<std::ptrdiff_t>(idx));
            continue;
        }
        const std::uint16_t typ = static_cast<std::uint16_t>(buf_[2] | (buf_[3] << 8));
        const std::size_t length =
            static_cast<std::size_t>(buf_[4]) | (static_cast<std::size_t>(buf_[5]) << 8) |
            (static_cast<std::size_t>(buf_[6]) << 16) | (static_cast<std::size_t>(buf_[7]) << 24);
        if (length > max_unit_) {
            buf_.erase(buf_.begin()); // garbage — shift
            continue;
        }
        const std::size_t total = 8 + length;
        if (buf_.size() < total)
            break; // wait for the rest (split between reads)
        Bytes payload(buf_.begin() + 8, buf_.begin() + static_cast<std::ptrdiff_t>(total));
        buf_.erase(buf_.begin(), buf_.begin() + static_cast<std::ptrdiff_t>(total));
        ++units_;
        type_counts_[typ] += 1;
        if (on_unit_)
            on_unit_(typ, payload);
        if (typ == COMPOSITE_TYPE_DUML && on_duml_) {
            on_duml_(payload);
        } else if ((typ == kVideoA || typ == kVideoD) && on_video_) {
            on_video_(payload);
        }
    }
}

void CompositeDemux::feed(const Bytes& data) {
    feed(data.data(), data.size());
}

} // namespace djilink
